src/core/sr.ml

```diff
@@ -102,50 +61,184 @@ let check_rule : sym StrMap.t -> sym * pp_hint * rule Pos.loc -> unit =
     | Wild        -> assert false (* Cannot appear in LHS. *)
     | TRef(_)     -> assert false (* Cannot appear in LHS. *)
   in
-  let lhs = List.map (fun p -> Bindlib.unbox (to_m 0 p)) r.elt.lhs in
-  let lhs = Basics.add_args (Symb(s,h)) lhs in
-  let metas = Array.map (function Some m -> m | None -> assert false) metas in
-  (* We substitute the RHS with the corresponding metavariables. *)
-  let fn m =
-    let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
-    let xs = Bindlib.new_mvar mkfree xs in
-    let e = Array.map _Vari xs in
-    TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m e)))
+  trans
+
+(** [new_function_symbol name a] creates a fresh function symbol named [name],
+    with type [a].  The symbols created using this function should not outlive
+    the SR-checking phase. *)
+let new_function_symbol name a =
+  { sym_name = name ; sym_type = ref a ; sym_path = [] ; sym_def = ref None
+  ; sym_impl = [] ; sym_rules = ref [] ; sym_tree = ref Tree_types.empty_dtree
+  ; sym_prop  = Defin ; sym_expo  = Public }
+
+(** Mapping of pattern variable names to their reserved index. *)
+type index_tbl = (string, int) Hashtbl.t
+
+(** [symb_to_tenv pr syms htbl t] builds a RHS for the pre-rule [pr]. The term
+    [t] is expected to be a version of the RHS of [pr] whose term environments
+    have been replaced by function symbols of [syms]. This function builds the
+    reverse transformation, replacing symbols by the term environment variable
+    they stand for.  Here, [htbl] should give the index in the RHS environment
+    for the symbols of [syms] that have corresponding [term_env] variable. The
+    pre-rule [pr] is provided to give access to these variables and also their
+    expected arities. *)
+let symb_to_tenv : pre_rule Pos.loc -> sym list -> index_tbl -> term -> tbox =
+    fun {elt={pr_vars=vars;pr_arities=arities;_};pos} syms htbl t ->
+  let rec symb_to_tenv t =
+    log_subj "symb_to_tenv %a" pp_term t;
+    let (h, ts) = Basics.get_args t in
+    let ts = List.map symb_to_tenv ts in
+    let (h, ts) =
+      match h with
+      | Symb(f,_) when List.memq f syms ->
+          let i =
+            try Hashtbl.find htbl f.sym_name with Not_found ->
+            fatal pos "Introduced symbol [%s] cannot be removed." f.sym_name
```

Comment:
I don't understand how this could happen. Please explain. Do you have an example?

Reply:
Do you at least agree that not all the symbols we introduce may correspond to a term environment variable? If yes, it seems obvious to be that such substituted symbol could escape. If no, can you explain why?

Reply:
Do you mean a symbol representing a metavariable occurring in the type of a pattern variable?

Reply:
Yes, that is what I have in mind.

Reply:
Ok. Then, for the record, thanks to add a comment.